src/backend/empty/src/memory.rs

```diff
@@ -0,0 +1,140 @@
+use hal::{
+    device::{AllocationError, MapError, OutOfMemory},
+    memory::Segment,
+    MemoryTypeId,
+};
+use std::cell::UnsafeCell;
+use std::convert::TryInto;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A memory allocation.
+#[derive(Debug)]
+pub struct Memory {
+    /// The type of memory this block was allocated from.
+    memory_type: MemoryTypeId,
+    /// Size of the allocated memory block.
+    size: u64,
+    /// The backing memory allocation.
+    data: UnsafeCell<Box<[u8]>>,
+    /// Marker for determining if data is currently mapped.
+    mapped: AtomicBool,
+}
+
+impl Memory {
+    pub fn allocate(memory_type: MemoryTypeId, size: u64) -> Result<Self, AllocationError> {
+        assert_eq!(memory_type.0, 0, "We only support one memory type");
+
+        // TODO: track how many allocations are made and set a limit
+
+        // Allocate a buffer in RAM
+        let data = {
+            let size = size
+                .try_into()
+                // If we're on 32-bit and the given size is greater than 2^32,
+                // we certainly can't allocate it.
+                .map_err(|_| AllocationError::OutOfMemory(OutOfMemory::Host))?;
+
+            vec![0u8; size].into_boxed_slice()
+        };
+
+        let memory = Memory {
+            memory_type,
+            size,
+            data: UnsafeCell::new(data),
+            mapped: AtomicBool::new(false),
+        };
+        Ok(memory)
+    }
+
+    pub fn map(&self, segment: Segment) -> Result<*mut u8, MapError> {
+        if segment.offset >= self.size {
+            return Err(MapError::OutOfBounds);
+        }
+        if let Some(size) = segment.size {
+            if segment.offset + size > self.size {
+                return Err(MapError::OutOfBounds);
+            }
+        }
+
+        let order = Ordering::SeqCst;
+        self.mapped
+            .compare_exchange(false, true, order, order)
```

Comment:
I wanted to ensure the memory isn't mapped twice so I tried using an `AtomicBool`. It *seems* to achieve the intended behavior?

Reply:
as said in another comment, we don't even need to check if it's mapped. We need to assume the user knows what they are doing

Reply:
Isnt it useful for a mock backend to check its usage?
A real backend will have vulkan validation layers (or similar) so this is not that different?
Or is it just not worth the complexity?

Reply:
You can't validate *all* usage, it's a huge task that we can't accomplish.
So if we are to draw the line, it would be very close towards "not validating anything".

It also is a question of what the backend will be used for. Previously, we only used it to check if something builds.
Next, we want to use it for gfx-extras testing. This doesn't require any validation, it basically just need to not panic.

Are you suggesting that people use the backend at some later point to validate their workflows?

Reply:
Honestly im not suggesting anything, I was just curious as to your reasoning here, and your response was enlightening so thankyou. :)